File: TemplateMethod/dart/Task.dart
import 'AuditTrail.dart';

abstract class Task {
  AuditTrail _auditTrail;

  Task({
    required auditTrail,
  }) : _auditTrail = auditTrail;

  void execute() {
    _auditTrail.record();
    doExecute();
  }

  void doExecute();
}

File: TemplateMethod/python/task.py
from abc import ABCMeta, abstractmethod
from audit_trail import AuditTrail


class Task(metaclass=ABCMeta):
    def __init__(self, audit_trail: AuditTrail) -> None:
        self.__audit_trail = audit_trail

    def execute(self) -> None:
        self.__audit_trail.record()
        self.doExecute()

    @abstractmethod
    def doExecute(self) -> None:
        pass
